_drafts/cortexm-and-clang.md

```diff
@@ -0,0 +1,971 @@
+$ tar -xf clang+llvm-10.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz
+```
+
+Let's add the toolchain to our path:
+
+```bash
+# OSX
+$ export PATH=`pwd`/clang+llvm-10.0.0-x86_64-apple-darwin/bin/:$PATH
+# Ubuntu
+$ export PATH=`pwd`/clang+llvm-10.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/:$PATH
+```
+
+You can sanity check that the install is working by running `clang --version`:
+
+```bash
+$ clang --version
+clang version 10.0.0
+Target: x86_64-apple-darwin18.5.0
+Thread model: posix
+InstalledDir: <SHOULD MATCH PATH YOU JUST ADDED ABOVE>
+```
+
+## Static Analysis With Clang and scan-build
+
+[Below](#porting-gcc-to-clang) we will dive into how to migrate a project using GCC to Clang, but
+first let’s take a look at some of the awesome static analysis passes and checkers we can run on
+our example project with LLVM/Clang!
+
+### Overview of scan-build
+
+`scan-build` is a static analyzer included as part of the LLVM toolchain. It works by intercepting calls to gcc or clang and making a static analysis pass.
+
+`scan-build` works by overriding the `CC` and `CXX` used by `CMake` and `Make` so if you aren't doing anything tricky in your project, invoking the tool is as simple as:
+
+```
+$ scan-build make
+$ scan-build cmake ..
+```
+
+`scan-build` has a number of "CHECKERS"[^7] enabled by default. You can view them by running `scan-build --help`:
+
+```bash
+$ scan-build --help
+CONTROLLING CHECKERS:
+
+ A default group of checkers are always run unless explicitly disabled.
+ Checkers may be enabled/disabled using the following options:
+
+ -enable-checker [checker name]
+ -disable-checker [checker name]
+[...]
+AVAILABLE CHECKERS:
+
+ + core.CallAndMessage           Check for logical errors for function calls and Objective-C message expressions (e.g., uninitialized arguments, null function pointers)
+ + core.DivideZero               Check for division by zero
+ + core.DynamicTypePropagation   Generate dynamic type information
+ + core.NonNullParamChecker      Check for null pointers passed as arguments to a function whose arguments are references or marked with the 'nonnull' attribute
+ + core.NullDereference          Check for dereferences of null pointers
+ + core.StackAddressEscape       Check that addresses to stack memory do not escape the function
+ + core.UndefinedBinaryOperatorResult Check for undefined results of binary operators
+[...]
+NOTE: "+" indicates that an analysis is enabled by default.
+```
+
+I find the default set to be a good starting point but you can easily enable non-default ones or
+disable individual defaults. (You can even implement your own custom pugins but these topics are outside the scope of this article.)
+
+Let's run scan-build on our example project. We'll see it finds a few interesting potential issues that neither Clang or GCC noticed during compilation.
+
+Results will be emitted directly to the console or you can browse them using the `scan-view`
+command bundled with LLVM.
+
```

Comment:
Nooo!!!! You *have* to include a screenshot of the `scan-view` website! 

> Imagine if Jenkins would publish the HTML of each build artifact and you can always view the current static analysis issues of the master branch!

Reply:
hehe I was thinking of making one of them a thumbnail but you are right including some here would be a good call too!